Add amdflang to CI

samhatfield opened the pull request.

samhatfield committed:
Add amdflang to CI

Co-authored-by: Paul Mullowney <Paul.Mullowney@amd.com>

samhatfield committed:
Add amdflang to CI

Co-authored-by: Paul Mullowney <Paul.Mullowney@amd.com>

samhatfield committed:
Disable MPI for amdclang

samhatfield committed:
Allow matrix CMake options to take precedence for overrides

samhatfield committed:
Add -fPIC option for amdflang

samhatfield committed:
Add -fPIC option for amdflang

samhatfield committed:
Add -fopenmp for amdclang in CI

samhatfield committed:
Add -fopenmp for amdclang in CI

samhatfield commented:
Status:
- MPI has been disabled for amdflang. Somehow the OpenMPI built by amdflang can't be found by CMake. We will work on this in a future PR.
- Otherwise all tests are passing with amdflang except for `ectrans_test_install ` which segfaults when running `transi_sptogp`. This is very similar to what we saw with IntelLLVM actually, which is very suspicious. However I cannot yet debug this crash interactively as I can't build a working amdflang on ECMWF's systems (I could probably build amdflang on my Mac but I have a feeling it's OS-specific). Will liaise with @PaulMullowney about how to debug this further.

samhatfield committed:
Increase stack size to unlimited for ectrans_test_install

samhatfield committed:
Increase stack size to unlimited for ectrans_test_install

samhatfield commented:
Note #249 must be merged into this in order for the amdflang tests to pass.

wdeconinck reviewed (approved):
This is great! Thanks also for the install-amd-flang script. Very useful.

marsdeno commented:
Excellent!

PaulMullowney commented:
We should hook the GPU build in, next. It will runtime fail, but at least we'll be covering compilation.

samhatfield commented:
> Note #249 must be merged into this in order for the amdflang tests to pass.

Correction: #249 must be merged into this in order for the amdflang tests to _run_. In this branch the `TESTS` feature is disabled for amdflang because we are building without MPI. After #249 is merged to develop (and I rebase this branch) we will be able to enable `TESTS` without MPI and we can then check whether the test suite passes for amdflang.

wdeconinck commented:
@samhatfield I rebased on develop.
Question. Why do we need to add the -fPIC and -fopenmp flags directly?
Does CMake not detect these flags automatically?
There exist alternatives like `-DOpenMP_Fortran_FLAGS=-fopenmp` to give CMake hints about which OpenMP Flag to use.

samhatfield commented:
> @samhatfield I rebased on develop. Question. Why do we need to add the -fPIC and -fopenmp flags directly? Does CMake not detect these flags automatically? There exist alternatives like `-DOpenMP_Fortran_FLAGS=-fopenmp` to give CMake hints about which OpenMP Flag to use.

Good question. It might be that I added those flags before I added the requirement for a CMake version that is amdflang-aware. I'll try removing them now and see what happens.

samhatfield committed:
Revert "Add -fopenmp for amdclang in CI"

This reverts commit 63839fc10182aeff77303fce6b11fb310da56cb9.

samhatfield committed:
Revert "Add -fopenmp for amdclang in CI"

This reverts commit 63839fc10182aeff77303fce6b11fb310da56cb9.

samhatfield committed:
Use OpenMP_{lang}_FLAGS

samhatfield committed:
Use OpenMP_{lang}_FLAGS

samhatfield commented:
- As for the OpenMP flag issue, I have tried using instead `OpenMP_{lang}_FLAGS=-fopenmp` but this isn't working. I would like to use a cleaner CMake-idiomatic solution but it's difficult to debug this as I can't reproduce it locally.

samhatfield commented:
> > * As for the OpenMP flag issue, I have tried using instead `OpenMP_{lang}_FLAGS=-fopenmp` but this isn't working. I would like to use a cleaner CMake-idiomatic solution but it's difficult to debug this as I can't reproduce it locally.
> 
> Can you come up with a list of issues and we'll discuss at Friday's meeting?

I can. It's just not really amdflang-new's fault necessarily, just CMake.

samhatfield committed:
Update amdflang to 7.0.5

samhatfield committed:
Update amdflang to 7.0.5

samhatfield committed:
Update reference to amdflang

amdflang-new has now become amdflang.

samhatfield committed:
Update reference to amdflang

amdflang-new has now become amdflang.

samhatfield commented:
Any ideas @PaulMullowney ?

PaulMullowney commented:
> Any ideas @PaulMullowney ?
Are these CPU and/or GPU builds?

PaulMullowney commented:
> CPU builds, on the GH virtual machines

For a local install, libomp.so is in something like:
/PATH/TO/rocm-afar-8248-drop-7.0.5/lib/llvm/lib/libomp.so

You could try adding -L/PATH/TO/rocm-afar-8248-drop-7.0.5/lib/llvm/lib/

I know it's not a solution but we can at least try it as a workaround.

samhatfield committed:
Add llvm/lib to library load path for amdflang

samhatfield committed:
Add llvm/lib to library load path for amdflang

samhatfield committed:
Remove references to "NEW"

samhatfield committed:
Remove references to "NEW"

samhatfield committed:
Hide progress bar

samhatfield committed:
Hide progress bar

samhatfield committed:
Sort out version selection

samhatfield committed:
Sort out version selection

samhatfield commented:
> > CPU builds, on the GH virtual machines
> 
> For a local install, libomp.so is in something like: /PATH/TO/rocm-afar-8248-drop-7.0.5/lib/llvm/lib/libomp.so
> 
> You could try adding -L/PATH/TO/rocm-afar-8248-drop-7.0.5/lib/llvm/lib/
> 
> I know it's not a solution but we can at least try it as a workaround.

I *think* the issue is that lib/llvm/lib is not on the `LD_LIBRARY_PATH` at the point when `ectrans-benchmark` is invoked. What you're suggesting would be the solution to a link-time failure, right? Whereas what we have is a runtime failure.

Still you gave me an idea. I tried adding an export of `LD_LIBRARY_PATH` in the install-amdflang step, but then it occurred to me that env vars are not by default shared between different steps of an action. So we would have to add that export to the `Build & Test` stage, but only for amdflang. Not sure how to do that. I'm also not sure why other compilers don't have this issue.

samhatfield committed:
Ensure LD_LIBRARY_PATH is on GITHUB_ENV

samhatfield committed:
Ensure LD_LIBRARY_PATH is on GITHUB_ENV

samhatfield commented:
amdflang is giving a segfault for the direct and inverse transform adjoint tests. I've spent a day trying to figure out what's going wrong, but sadly none the wiser. @PaulMullowney thinks it might be a stacksize issue, but as far as I can tell there's no way to modify the ulimits in our reusable action. I've tried simplifying the dirtrans test by removing the `SCALPRODGP` function and just directly calculating the values in the main program, reusing already allocated arrays as much as possible. But somehow, the segfault still creeps back in. I'll pick this up again later.

samhatfield commented:
After offline discussion with @reuterbal and @PaulMullowney, we strongly suspect that amdflang is more stack hungry than other compilers, and that the segfaults are due to running out of stack space. Unfortunately we don't believe it's possible to increase the stack size given the restricted structure of the reusable action we're using to build and test ecTrans.

Assuming this theory is correct, we have three options:
- Disable the failing tests for amdflang.
- Reduce the resolution of the failing tests to reduce the memory consumption. This is what I've done in this PR. I've also simplified the failing tests to reduce the size of allocated memory (mostly reusing work arrays).
- Try and write a wrapper script, akin to [test-install.sh.in](https://github.com/ecmwf-ifs/ectrans/blob/develop/tests/test-install.sh.in) which calls `ulimit -s unlimited`.

wdeconinck commented:
@samhatfield I tried installing latest amd-flang drop 22.3.0 and I noticed the flag `-fno-stack-arrays` exists.
It could be worth trying?
Also your install script does not work with just this latest drop 22.3.0 ; that should be adapted.

samhatfield committed:
Update amdflang version

samhatfield committed:
Update amdflang version

samhatfield committed:
Show progress when downloading

samhatfield committed:
Show progress when downloading

samhatfield commented:
I'll try upgrading to amdflang 22.2.0 first, then adding -fno-stack-arrays.

samhatfield committed:
Update amdflang to 22.2.0

samhatfield committed:
Update amdflang to 22.2.0

samhatfield committed:
Revert "Show progress when downloading"

This reverts commit 093c6f6ebf1d57bced40267f8f464644f04a368c.

TOO MUCH OUTPUT.

samhatfield committed:
Revert "Show progress when downloading"

This reverts commit 093c6f6ebf1d57bced40267f8f464644f04a368c.

TOO MUCH OUTPUT.

samhatfield committed:
Add heap arrays option

samhatfield committed:
Add heap arrays option

samhatfield committed:
Try semicolons

samhatfield committed:
Try semicolons

samhatfield committed:
(debug)

samhatfield committed:
(debug)

samhatfield committed:
(debug)

samhatfield committed:
(debug)

samhatfield committed:
Try using arch files to pass flags

samhatfield commented:
Now compiling with -fno-stack-arrays. Still doesn't work :(

samhatfield committed:
Allow matrix CMake options to take precedence for overrides

samhatfield committed:
Disable MPI for amdclang

msleigh reviewed (approved):
Approving from copyright point of view only
